packages/flutter/lib/src/widgets/platform_menu_bar.dart

```diff
@@ -0,0 +1,886 @@
+/// [MenuItem]s, not a tree of widgets.
+///
+/// {@tool sample}
+/// This example shows a [PlatformMenuBar] that contains a single top level
+/// menu, containing three items for "About", a toggleable menu item for showing
+/// a message, a cascading submenu with message choices, and "Quit".
+///
+/// **This example will only work on macOS.**
+///
+/// ** See code in examples/api/lib/material/platform_menu_bar/platform_menu_bar.0.dart **
+/// {@end-tool}
+class PlatformMenuBar extends StatefulWidget with DiagnosticableTreeMixin {
+  /// Creates a const [PlatformMenuBar].
+  ///
+  /// The [body] and [children] attributes are required.
+  const PlatformMenuBar({
+    Key? key,
+    required this.body,
+    required this.children,
+  }) : super(key: key);
+
+  /// The widget to be rendered in the Flutter window that these platform menus
+  /// are associated with.
+  ///
+  /// This is typically the body of the application's UI.
+  final Widget body;
+
+  /// The list of menu items that are the top level children of the
+  /// [PlatformMenuBar].
+  ///
+  /// The `children` member contains [MenuItem]s. They will not be part
+  /// of the widget tree, since they are not widgets. They are provided to
+  /// configure the properties of the menus on the platform menu bar.
+  ///
+  /// Also, a Widget in Flutter is immutable, so directly modifying the
+  /// `children` with `List` APIs such as
+  /// `somePlatformMenuBarWidget.children.add(...)` will result in incorrect
+  /// behaviors. Whenever the children list is modified, a new list object
+  /// should be provided.
+  final List<MenuItem> children;
+
+  @override
+  State<PlatformMenuBar> createState() => _PlatformMenuBarState();
+
+  @override
+  List<DiagnosticsNode> debugDescribeChildren() {
+    return children.map<DiagnosticsNode>((MenuItem child) => child.toDiagnosticsNode()).toList();
+  }
+}
+
+class _PlatformMenuBarState extends State<PlatformMenuBar> {
+  List<MenuItem> descendants = <MenuItem>[];
+
+  @override
+  void initState() {
+    super.initState();
+    assert(
+        WidgetsBinding.instance.platformMenuDelegate.debugLockDelegate(context),
+        'More than one active $PlatformMenuBar detected. Only one active '
+        'platform-rendered menu bar is allowed at a time.');
+    WidgetsBinding.instance.platformMenuDelegate.clearMenus();
+    _updateMenu();
+  }
+
+  @override
+  void dispose() {
+    assert(WidgetsBinding.instance.platformMenuDelegate.debugUnlockDelegate(context),
+        'tried to unlock the $DefaultPlatformMenuDelegate more than once with context $context.');
+    WidgetsBinding.instance.platformMenuDelegate.clearMenus();
+    super.dispose();
+  }
+
+  @override
+  void didUpdateWidget(PlatformMenuBar oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    final List<MenuItem> newDescendants = <MenuItem>[
+      for (final MenuItem item in widget.children) ...<MenuItem>[
+        item,
+        if (item is PlatformMenu) ...item.descendants,
```

Comment:
can this be done polymorphic without type checking by calling a general interface API on MenuItem? Leaf items could just return null to indicate that they don't have descendants.